### DISCUSSION TITLE
Welcome to batch Discussions!

### DISCUSSION THREAD

**Discussion by jbcaillau** [pns-mam/batch#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jbcaillau** [comment]
@laitlito @nathanedr Ça se passe ici pour les discussions.

**Comment by laitlito** [comment]
Bonjour monsieur,
J'ai remis au propre les notes que j'ai prises lors de notre discussion de ce matin et j'aimerai vous demander si j'ai bien mis au propre le problème : 
De ce que j'ai compris, on a deux fonctions $F_0$ et $F_1$ toutes deux de $\mathbb{R}^4 \longrightarrow \mathbb{R}^4$, que l'on trouve dans le code du repository.
On calcule $F_{01} = \[F_0, F_1\](\varphi) = \nabla F_1 \times F_0 - \nabla F_0 \times F_1$ où $\nabla$ est la jacobienne.
Ensuite, on calcule $F_{101} = \[F_1, F_{01}\] = \nabla F_{01}\times F_1 - \nabla F_1 \times F_{01}$
Avec ces deux crochets calculés, on doit vérifier
$$rang(F_1, F_{01}) = rang(F_1,F_{01}, F_{101}) = 2$$ et ce, pour tout $\varphi$ et pour toutes constantes $k$ et $K$
Est-ce bien ça pour la partie maths ?
Bonne soirée
Lélio

**Comment by jbcaillau** [comment]
# Point du 9/10/2023

Bonsoir à tous les deux ; @laitlito oui, c'est OK. Pour résumer : 
- [x] 1. calcul de rang
- [x] 2. résolution numérique

## 1. Calcul de rang

Mettre le système

![IMG_2623](https://github.com/pns-mam/batch/assets/62183989/eb34e8c5-104a-4566-a5fa-686bbf3aad1d)

sous la forme

$$ \dot{\varphi} = F_0(\varphi) + u F_1(\varphi) $$

avec $F_0$ et $F_1$ les deux champs de vecteurs (dépendant des $4$ paramètres $k_R$, $k_M$, $K_R$, $K_M$) déjà codés dans [batch.jl](https://github.com/pns-mam/batch/blob/main/code/batch.jl). On veut montrer que, quel que soit $\varphi$ dans $\mathbf{R}^4$ (et quels que soient les paramètres cités), (i) la famille $\lbrace F_1,F_{01} \rbrace$ est de rang $2$, (ii) $F_{101}$ appartient au sev engendré par ces deux vecteurs. 

Le [crochet de Lie](https://en.wikipedia.org/wiki/Lie_bracket_of_vector_fields) de deux champs de vecteurs [^1] $F$ et $G$ (on note $x \in \mathbf{R}^n$ la variable est défini par [^2]

$$ [F,G] (x) = G'(x)\cdot F(x) - F'(x)\cdot G(x). $$

## 2. Résolution numérique

Il s'agit de reproduire en Julia les simulations numériques de la section 4.2 du [preprint](xxxx) sur la maximisation du volume à temps final fixé ("horizon fini"). Utiliser pour cela le package [OptimalControl.jl](https://control-toolbox.org/OptimalControl.jl/stable) de [control-toolbox](https://control-toolbox.org).
 
[^1]: Un champ de vecteurs $F$ est une fonction (dérivable...) de $\mathbf{R}^n$ dans lui-même.

[^2]: Ne pas noter $\nabla$ pour la dérivée / matrice jacobienne, ce symbole est réservé au gradient.

**Comment by jbcaillau** [comment]
@laitlito @nathanedr Bonsoir à tous les deux, rdv ce mercredi 17:15 à mon bureau si OK pour vous. (On s'arrête à 18:00 🤞🏾) J'attend votre confirmation.

**Reply by laitlito** [reply]
Bonjour monsieur,
C'est noté, on se retrouve mercredi à 17h15
Bonne journée

**Comment by jbcaillau** [comment]
# Point du 18/10/2023
@laitlito @nathanedr 
- pour les calculs de rang, et tout d'abord de crochets de Lie, passer en calcul formel ("symbolique") à l'aide du package [Symbolics.jl](https://symbolics.juliasymbolics.org/stable/manual/derivatives/)
- en parallèle, pour la résolution numérique du problème de commande optimale, voir [cet exemple](https://control-toolbox.org/docs/optimalcontrol/stable/tutorial-goddard.html) qui est assez proche (système affine en le contrôle, de la forme $\dot{x}(t) = F_0(x(t)) + u(t) F_1(x(t))$ )

**Comment by jbcaillau** [comment]
@laitlito @nathanedr rdv mercredi (25/10) 17:25 ?

**Reply by laitlito** [reply]
Bonjour monsieur,
C'est ok pour nous ce mercredi
Nous essayons actuellement de travailler sur le projet avec les sessions étudiantes mais Julia n'est pas installé... 
L'ordinateur de Nathan est très lent et le miens est encore en réparation (je devrais le récupérer dans la semaine)
Bonne journée

**Reply by jbcaillau** [reply]
Re ; téléchargez sur vos comptes à l'école (= en local, c'est pas gros) la dernière stable release :
https://julialang.org/downloads/

Par ex. par la commande : 
```unix
curl -L -O https://julialang-s3.julialang.org/bin/linux/x64/1.9/julia-1.9.3-linux-x86_64.tar.gz
```

**Reply by agustinyabo** [reply]
Vous faites en visio ou en vie réelle ? Je peux participer ?

**Reply by jbcaillau** [reply]
@agustinyabo Pour toi on fera les deux : rdv [zoom](https://univ-cotedazur.zoom.us/my/caillau) demain 17:15. (@laitlito @nathanedr et à mon bureau pour vous)

**Comment by jbcaillau** [comment]
# Point du 25/10/2023
@laitlito @nathanedr @agustinyabo 
- pour le calcul des dérivées en formel (= calcul symbolique), si pas de calcul de dérivée (jacobienne) pour une fonction à valeurs vectorielles, $f : \mathbf{R}^n \to \mathbf{R}^m$, utiliser un calcul de gradient coordonnée par coordonnée 👍 

$$ f'(x) = \left[ \begin{array}{c}
  \dots \ ^t \nabla f_1(x) \dots\\
  \vdots\\
  \dots \ ^t \nabla f_m(x) \dots
\end{array} \right]. $$

- si les expressions formelles ne se simplifient pas bien, penser à évaluer en une valeur particulière pour voir si le rang, une structure particulière... apparaissent
- dans la définition (`@def...`) du problème de contrôle optimal, ajouter les contraintes sur l'état sur $\varphi = (s,p,r,V)$ décrites ci-dessous (ça aide l'optimiseur à limiter sa zone de recherche de solution) : 
![IMG_2664](https://github.com/pns-mam/batch/assets/62183989/f8f5d37e-300a-42f8-86be-944a00b5fbcb)
- les conditions initiales pour les résolutions à reproduite sont données dans les légendes des graphes des simulations ($s_0=...$)
- prochain point (présentiel + [zoom](https://univ-cotedazur.zoom.us/my/caillau)) le 8/11 17:15.

**Comment by laitlito** [comment]
Bonsoir, 
Nous avons mis sur le repository le début de code que nous avons écrit. Pour l'OCP, nous avons du mal à comprendre comment analyser les graphes. Quant au calcul du rang, il semble qu'il y ait une erreur vu qu'on trouve que le rang de $[F_1\ F_{01}] = 4$ ce qui n'est pas possible il me semble. Avez-vous une idée de pourquoi ?
Bonne soirée

**Reply by jbcaillau** [reply]
@laitlito @nathanedr Bonjour à tous les deux ; je regarde le code et reviens vers vous. Réponse immédiate quant au rang : une matrice $4 \times 2$ ne peut définitivement pas être de rang $4$ (rang max = $2$, ce qu'on souhaite vérifier ici) !

**Reply by jbcaillau** [reply]
@laitlito @nathanedr Bien vu ce [code](https://github.com/pns-mam/batch/blob/main/code/finiteTimeHorizonOptimalConrolProblem.jl) qui génère les résultats ci-dessous : 
- l'idée est de comparer vos résultats à ceux dans le papier Fig. 4
- dans l'immédiat, le principal pb est que vous avez oublié la contraint $0 \leq u(t) \leq 1$ sur le contrôle... d'où les valeurs en $1e15$ pour celui-ci ; j'ai poussé une correction du code, je vous laisse vérifier, tester et comparer
- vous avez `grid_size=100`, essayez `200, 300, 400, 500...` et dites nous svp si ça *converge* ou pas (résultats de l'appel à `Ipopt` via `solve`)

![IMG_2682](https://github.com/pns-mam/batch/assets/62183989/57304a99-7f44-483d-b434-ac9e51ca4a55)

**Comment by jbcaillau** [comment]
@laitlito @nathanedr @agustinyabo re-bonjour à tous les deux, je dois reporter notre point projet de ce soir, désolé pour ça. Je vous propose un point en visio mardi 14/11 à 17:45 : dites moi si OK pour vous.

**Reply by nathanedr** [reply]
Oui c’est bon pour nous

**Reply by laitlito** [reply]
Bonjour monsieur,
Je vous informe que je ne pourrais malheureusement pas être présent ce soir. Nous avons vu avec Nathan et il m'a dit qu'il me ferait un résumé de la visio
Bonne journée

**Reply by jbcaillau** [reply]
@laitlito @nathanedr @agustinyabo Bonjour tous les deux, rdv 17:45 sur ce [zoom](https://univ-cotedazur.zoom.us/my/caillau)

**Comment by jbcaillau** [comment]
# Point du 14/11/2023
@laitlito @nathanedr @agustinyabo 
- [x] pousser la vérif. formelle du rang $2$ pour $F_1$, $F_{01}$ (il faut et il suffit qu'il existe un bloc $2 \times 2$ inversible)
- [x] idem pour le fait que le rang de $F_1$, $F_{01}$, $F_{101}$ reste $2$ (vérifier cette fois que les $4$ sous-matrices $3 \times 3$ sont de déterminant *nul*)
- [x] pousser le code de résolution corrigé après comparaison avec la Fig. 4
- prochain rdv : lundi 27/11 15:45

**Comment by jbcaillau** [comment]
# Point du 27/11/2023
@laitlito @nathanedr @agustinyabo 
- finaliser les points ci-avant
- pour le [test en symbolique](https://github.com/pns-mam/batch/blob/main/code/LieBracketsRankVerification.ipynb), les mineurs $2 \times 2$ associés à $F_1$, $F_{01}$ semblent bien non nuls (il suffit qu'un seul le soit pour assurer le rang $2$), *cf.* positivité des états $s$, $p$, $\mathcal{V}$...
- pour les mineurs $3 \times 3$ (TBP [^1]), on s'attend à ce que les $4 = C^3_4$ soient nuls : dans la mesure où ça semble ne pas se simplifier, évaluer numériquement pour essayer de devenir ce qu'il en est (et localiser les zéros éventuels)
- dans les deux cas, tester aussi en symbolique en conservant des fonctions arbitraires pour $w_R$ et $w_M$
- vérifier la date de remise du rapport et les attendus (format, taille...)
- soutenance 20/12 ou 21/12
- pour connection `github` installer le client [gh](https://github.com/cli/cli)
- prochain rdv : lundi 4/12 15:45 (@agustinyabo bien noté que tu es au Pérou - ¡suerte!)

[^1]: To Be Pushed 🙂

**Comment by jbcaillau** [comment]
# Point du 4/12/2023
@laitlito @nathanedr @agustinyabo 
- calcul (symbolique) de rang OK : l'évaluation numérique semble confirmer que les 4 mineurs $3 \times 3$ sont bien nuls
- [x] ça semble se lire à peu près directement sur les expressions non-simplifiées du [code](https://github.com/pns-mam/batch/blob/main/code/testRang.ipynb) : le vérifier
- [ ] simulation numérique : en l'état, solution de ce [code](https://github.com/pns-mam/batch/blob/main/code/resOptimal.jl) assez différente de la Fig. 4 du preprint : à creuser ;
- [ ] faire aussi les tests correspondant aux Fig. 5 et 6 (déterminer $\Sigma$, $\mathcal{V}_e$ et $\beta$ pour comparer les valeurs des coûts)
- [x] @agustinyabo peux-tu indiquer où trouver les valeurs de $\mathcal{V}_e$ et $\beta$ utilisées ? (OK pour $\Sigma$ qui ne dépend que des conditions initiales.)
- prochain point lundi 11/12 17:15

**Reply by agustinyabo** [reply]
Hola, JB. J'ai pas accès au papier sur Overleaf, tu pourrais me repartager ? De mémoire, $\beta = 0.003$, et $\mathcal{V}$ peut être rescaled without loss of generality, donc vous pouvez fixer $\mathcal{V}_e = 1$.

**Reply by jbcaillau** [reply]
@agustinyabo [Lien en édition](https://overleaf.irisa.fr/6865847839rkswykkryzkh)

**Comment by jbcaillau** [comment]
# Point du 11/12/2023
@laitlito @nathanedr @agustinyabo 
- [ ] tester simplification éventuelle avec autres fonctions $w_M$, $w_R$ (prendre des polynômes de degré au moins un, *etc*) ; si les résultats sont positifs, essayer un calcul symbolique avec des fonctions arbitraires (vérifier dans la doc de [Symbolics.jl](https://symbolics.juliasymbolics.org/stable) si c'est faisable)
- [x] finaliser les tests numériques sur la résolution du pb indiqués [ci-avant](https://github.com/pns-mam/batch/discussions/1#discussioncomment-7756028) (en l'état, pas de correspondance avec Fig. 4, 5 ou 6)
- pour le rapport : liens pour [crochet de Lie](https://en.wikipedia.org/wiki/Lie_bracket_of_vector_fields#In_coordinates), [intro. commande optimale](https://en.wikipedia.org/wiki/Optimal_control#General_method) ; prévoir aussi un mot rapide sur la méthode de travail (repo `github`)
- prochain point : 19/12/2023 14:00, répét. présentation (15' = 15 slides $\pm 2$)

**Comment by nathanedr** [comment]
Bonjour monsieur, on vient de pousser un nouveau code : testRangDiff.ipynb. On a réussi à implémenter $w_r$ et $w_m$ sans les déclarer en fonction de $K_r$, $k_r$, ... 
C'est tout de suite moins chargé mais il y a toujours le même problème pour verifier le rang de [F1 F01 F101] (tout ne s'annule pas)

**Comment by laitlito** [comment]
Bonjour monsieur, 
On a regardé les graphes du contrôle pour des $t_f$ différents de $30$ et on observe des choses intéressantes : 
Pour $t_f=30$, le graphe ne ressemble en rien à celui trouvé par @agustinyabo 
Pour $t_f=90$, on a bien les bang-bang de $0$ à $15$ puis de $60$ à $90$ et le contrôle $\approx 0.25$. Cependant, ce n'est pas du tout le cas pour le $t_f=30$ ce qui est assez étonnant et nous ne voyons toujours pas quoi modifier dans le code pour que cela marche correctement.
![tf=90](https://github.com/pns-mam/batch/assets/71383434/fb646c5d-bed3-409c-90b2-59c88227563e)
Enfin, pour $t_f=300$, on perd les bang-bang ainsi que le contrôle stable et le graphe de $p$ ne correspond pas non plus à ce qu'on aimerai avoir.
![tf=300](https://github.com/pns-mam/batch/assets/71383434/a67b9b2c-30cc-4e61-9167-29ea57e826f1)

**Comment by jbcaillau** [comment]
@laitlito @nathanedr @agustinyabo Bonsoir à tous les deux, merci pour le dernier push : 
- c'est bien pour les simus, avec $t_f=90$ le résultat est tout à fait cohérent
- pour les autres valeurs de $t_f$, il faudrait comparer les volumes obtenus avec les solutions du preprint, mais la structure du contrôle ne semble pas correcte (ça peut être un optimum local strict)
- pour le calcul symbolique, job done 👍🏽 ; je note qu'en l'état, pas d'obstruction à ce que la propriété de rang soit vraie pour des fonctions $w_x$ arbitraires : *to be continued*
- bien vu le rapport, c'est OK : poussez svp le source $\LaTeX$ sur le repo (et pas simplement le pdf), je ferai qques corrections très mineures (controll-toolbox -> control-toolbox 😬)
- idem pour les slides ; on en parle demain lors de la répet. (rdv 14:00)

**Reply by jbcaillau** [reply]
PS. @laitlito @nathanedr Slides annotés joints
[SlidesAstrucEdery.pdf](https://github.com/pns-mam/batch/files/13716679/SlidesAstrucEdery.pdf)

**Comment by laitlito** [comment]
Bonjour monsieur,
J'ai push le code source du rapport, et voici le lien de partage [overleaf](https://fr.overleaf.com/4647868233jcbvfygdrhbv#fc1025) du diaporama.
Bonne soirée

**Comment by jbcaillau** [comment]
@laitlito @nathanedr merci à tous les deux pour le travail et bravo pour la [prés.](https://github.com/pns-mam/batch/blob/main/slides-astruc-edery.pdf). ~~Envoyez nous svp le lien (overleaf...) vers le rapport.~~ Dans le rapport, ajoutez svp les résultats de vos calculs avec $w_R$ et $w_M$.

@agustinyabo jette un œil sur [le rapport](https://github.com/pns-mam/batch/blob/main/Rapport%20batch/report.pdf), les jeunes ont montré que la structure (rang des crochets) est valable pour des fonctions arbitraires ($w_R$ et $w_M$).

**Reply by agustinyabo** [reply]
Trop bien, je vais regarder ça !

**Reply by jbcaillau** [reply]
🎄bonnes fêtes à tous !

**Reply by nathanedr** [reply]
Merci à vous aussi !

**Reply by laitlito** [reply]
Merci bonnes fêtes à vous aussi

**Reply by agustinyabo** [reply]
J'ai lu le rapport, très bien ! C'est une très bonne contribution. Petit détail : je changerais juste "PDE" par "ODE" à chaque fois.

Bonnes fêtes de l'autre bout du monde !

**Reply by agustinyabo** [reply]
D'ailleurs... très drôle l'idée qui les IA ont des bioréacteurs. J'avoue que dans la vraie vie c'est beaucoup moins stylé.

**Reply by jbcaillau** [reply]
@laitlito @nathanedr merci @agustinyabo pour la relecture : j'ai bien changé PDE en ODE dans le rapport, mais pb de compilation avec `xelatex` chez moi (fontes manquantes) ; du coup j'ai remis l'ancien pdf. Nathan et Lelio, pouvez-vous svp recompiler le rapport et le pousser ? Si possible après avoir inclus vos derniers résultats (calcul de rang avec fonctions arbitraires) ?

**Reply by nathanedr** [reply]
Très bien on vous fait cela !

**Reply by laitlito** [reply]
Bonjour, 
Le push du rapport avec l'ajout des résultats pour n'importe quelles fonction arbitraire + le code en annexe est fait
Bonne soirée

**Reply by jbcaillau** [reply]
@laitlito @nathanedr 🙏🏽 Grand merci à tous les deux, bien vu le rapport. Remarques rapides :  
- il semble manquer une parenthèse fermante dans le 4eme mineur $2 \times 2$ p. 13
- toujours p. 13, c'est étonnant qu'il n'y ait ni dérivée de $w_M$ dans $F_{01}$, ni dérivée d'ordre supérieur de $w_R$ dans $F_{101}$ (*cf.* terme $\beta$)

En tout cas, on va reprendre vos calculs qui nous ont bien aidé 👍🏽

**Reply by laitlito** [reply]
Bonjour monsieur, nous avons push la modification sur le mineur p. 13.
Pour le $\beta$, c'est bien comme ça dans notre code, à voir si on s'était trompé quelque part.
Bonne journée

**Reply by jbcaillau** [reply]
@laitlito @nathanedr bien reçu, merci !
